src/smtencoding/FStar.SMTEncoding.Encode.fs

```diff
@@ -370,28 +365,82 @@ let encode_free_var uninterpreted env fv tt t_norm quals =
               let guard, decls1 = match pre_opt with
                 | None -> mk_and_l guards, decls1
                 | Some p -> let g, ds = encode_formula p env' in mk_and_l (g::guards), decls1@ds in
-              let vtok_app = mk_Apply vtok_tm vars in
+              let dummy_var = mk_fv ("@dummy", dummy_sort) in
+              let dummy_tm = Term.mkFreeV dummy_var Range.dummyRange in
+              let should_thunk =
+                let is_type t =
+                    match (SS.compress t).n with
+                    | Tm_type _ -> true
+                    | _ -> false
+                in
+                let is_squash t =
+                    let head, _ = U.head_and_args t in
+                    match (U.un_uinst head).n with
+                    | Tm_fvar fv ->
+                      Syntax.fv_eq_lid fv FStar.Parser.Const.squash_lid
+
+                    | Tm_refine({sort={n=Tm_fvar fv}}, _) ->
+                      Syntax.fv_eq_lid fv FStar.Parser.Const.unit_lid
 
-              let vapp = mkApp(vname, List.map mkFreeV vars) in //arity ok, see decl below, arity is |formals| = |vars| (#1383)
+                    | _ -> false
+                in
+                //Do not thunk ...
+                lid.nsstr <> "Prims"  //things in prims
+                && not (quals |> List.contains Logic) //logic qualified terms
+                && not (is_squash t_norm) //ambient squashed properties
+                && not (is_type t_norm) // : Type terms, since ambient typing hypotheses for these are cheap
+              in
+              let thunked, vars =
+                 match vars with
+                 | [] when should_thunk ->
+                   true, [dummy_var]
+                 | _ -> false, vars
+              in
+              let arity = List.length formals in
+              let vname, vtok_opt, env = new_term_constant_and_tok_from_lid_maybe_thunked env lid arity thunked in
+              let get_vtok () = Option.get vtok_opt in
+              let vtok_tm =
+                    match formals with
+                    | [] when not thunked -> mkFreeV <| mk_fv (vname, Term_sort)
+                    | [] when thunked -> mkApp(vname, [dummy_tm])
+                    | _ -> mkApp(get_vtok(), []) //not thunked
+              in
+              let vtok_app = mk_Apply vtok_tm vars in
+              let vapp = mkApp(vname, List.map mkFreeV vars) in //arity ok, see decl below, arity is |vars| (#1383)
```

Comment:
This whole region of code is the main part of the new functionality. Probably worth a review here @aseemr or @mtzguido

Reply:
Looks good to me! Nitpicking: I would thunk `should_thunk` :).

Also the explanation on this PR is really useful. Maybe we could put it within the file too as a "note", in the style of GHC (https://ghc.haskell.org/trac/ghc/wiki/Commentary/CodingStyle#Commentsinthesourcecode). But this is good to merge as-is IMO.